compiler/optimizer/VPHandlers.cpp

```diff
@@ -4241,13 +4241,65 @@ TR::Node *constrainNew(OMR::ValuePropagation *vp, TR::Node *node)
          vp->addGlobalConstraint(node, classConstraint->asClass()->getClassType());
       else
          vp->addGlobalConstraint(node, classConstraint);
+
+      // If it's an abstract class or an interface, the allocation cannot be removed
+      // because InstantiationError could be thrown
+      TR_OpaqueClassBlock *clazz = classConstraint->getClassType() ? classConstraint->getClassType()->getClass() : NULL;
+      if (clazz && TR::Compiler->cls.isConcreteClass(vp->comp(), clazz))
+         {
+         node->setAllocationCanBeRemoved(true);
+         }
       }
 
-   if (classConstraint && classConstraint->getClassType() &&
-      classConstraint->getClassType()->getClass())
+   vp->addGlobalConstraint(node, TR::VPNonNullObject::create(vp));
+
+   node->setIsNonNull(true);
+
+   return node;
+   }
+
+TR::Node *constrainNewvalue(OMR::ValuePropagation *vp, TR::Node *node)
+   {
+   constrainChildren(vp, node);
+
+   vp->createExceptionEdgeConstraints(TR::Block::CanCatchNewvalue, NULL, node);
```

Comment:
Can you please elaborate on why we need `CanCatchNewvalue` ? I ask, to understand why some sort of generic `CanCatchAllocation` would not have been sufficient to handle the different cases of `New`, `Newvalue`, `ArrayNew` etc.

Reply:
There are exceptions for `CanCatchArrayNew` and `CanCatchNewvalue` that don't apply to `CanCatchNew`.

- `CanCatchNewvalue` can catch `IllegalAccessError`.  `newvalue` is the opcode for `aconst_init` and `withfield`. Both java instructions could throw exception `IllegalAccessError` [1]
- `CanCatchArrayNew` can catch `"Exception"`, `"RuntimeException"`, `"IllegalAccessError"`, and `"NegativeArraySizeException"`[2]


[1]
>Otherwise, if the referenced value class does not belong to the same nest as the current class, any of the exceptions pertaining to the nestmate test ([5.4.4](https://docs.oracle.com/javase/specs/jvms/se16/html/jvms-5.html#jvms-5.4.4)) may be thrown.

>5.4.4. Access Control
...
If C is not accessible to D, then access control throws an IllegalAccessError. Otherwise, access control succeeds.

[2]
https://github.com/eclipse/omr/blob/d6867e8b6cfd6e36048a46a35b23ff970ff02ee1/compiler/il/OMRBlock.cpp#L1927-L1953